Closure_Project/rest_api/models.py

```diff
@@ -162,14 +180,84 @@ def __str__(self):
+        required_by_type = {k: set() for k in REQUIRED_COURSE_TYPES}
+        required_courses = set()
+
+        for group in groups:
+            group_courses = list(group.courses.all())
+
+            required_by_type[group.course_type].update(group_courses)
+            required_courses.update(group_courses)
+
+        done = {t: 0 for t in ALL_COURSE_TYPES}
+        counted = set()
+
+        for take in self.take_set.all():
+            course = take.course
+            if course in counted:
+                continue
+            counted.add(course)
+
+            if course.is_corner_stone:
+                done[CourseType.CORNER_STONE] += course.points
+
+            elif course in required_courses:
+                for course_type in REQUIRED_COURSE_TYPES:  # ordered by importance
+                    if course in required_by_type[course_type]:
+                        done[course_type] += course.points
+                        break  # prevents courses from being counted in more than one type
+
+            else:
+                done[CourseType.SUPPLEMENTARY] += course.points
+
+        result = {CourseType.MUST.name:
+                      {'required': track.points_must,
+                       'done': done[CourseType.MUST]},
+
+                  CourseType.FROM_LIST.name:
+                      {'required': track.points_from_list,
+                       'done': done[CourseType.FROM_LIST]},
+
+                  CourseType.CHOICE.name:
+                      {'required': track.points_choice,
+                       'done': done[CourseType.CHOICE]},
+
+                  CourseType.CORNER_STONE.name:
+                      {'required': track.points_corner_stones,
+                       'done': done[CourseType.CORNER_STONE]},
+
+                  CourseType.SUPPLEMENTARY.name:
+                      {'required': track.points_complementary,
+                       'done': done[CourseType.SUPPLEMENTARY]}}
+
+        return result
+
 
 class Take(models.Model):
     student = models.ForeignKey(Student, on_delete=models.CASCADE)
-    course = models.ForeignKey(Course, on_delete=models.CASCADE)
+    course = models.ForeignKey(Course, on_delete=models.CASCADE, related_name='course_id_1')
     year_in_studies = models.IntegerField(choices=Year.choices)
     semester = models.CharField(choices=Semester.choices, max_length=10)
 
     def __str__(self):
         return ', '.join((f'{self.course.course_id}',
                           f'year={self.year_in_studies}',
                           f'semester={self.semester.lower()}'))
+
+    @property
+    def type(self) -> CourseType:
+        if self.course.is_corner_stone:
+            return CourseType.CORNER_STONE
+
+        cg_set = self.student.track.coursegroup_set
+        matching_cg = cg_set.filter(courses__id=self.course.id).all()
+        types = {cg.course_type for cg in matching_cg}
+        for course_type in [CourseType.MUST, CourseType.FROM_LIST, CourseType.CHOICE]:
```

Comment:
Perhaps we could save a constant list of `REQUIRED_TYPES` somewhere? we use it both here and on `remaining`

Reply:
We definitely should refactor and save this list in Take class only.
Changed to use take.type instead of calculating it again